### ISSUE TITLE
Making it easier for users to get started with CTranslate2 models

### ISSUE THREAD

**Issue by jordimas** [OpenNMT/CTranslate2#1950]
**Problem**
Currently, users must manually convert models before they can get started. This extra step creates unnecessary friction and makes onboarding more difficult than it needs to be.

**Tasks**
- Create an organization in HuggingFace and add the CTranslate2 developers
- Create a pipeline that automatically uploads to HuggingFace all the models supported converted to CTranslate2
- Add  huggingface_hub as an optional dependency to the Python bindings and provide a mechanism to load models from HuggingFace using a prefix like hf://

Example:
`generator = ctranslate2.Generator("hf://ctranslate2/llama3-8b", device="cpu")`

If somebody wants to get this task, it will be great

**Comment by jncraton** [comment]
Thank you for sharing these thoughts. Something like you've described sounds like it could be very beneficial for on-boarding new users! I do have a few ideas about the path forward.

Huggingface currently lists [over 500 models that are finetunes of just Qwen 2.5 0.5b](https://huggingface.co/models?other=base_model:finetune:Qwen/Qwen2.5-0.5B). Supporting **everything compatible** would be a very large task, and someone would likely have to manage determining which models are worth fully supporting via the CTranslate2 HF organization.

I'm not sure that we need new functionality within the CTranslate2 codebase in order to make getting started simpler. A few well-tested example models within a CTranslate2 HF organization may be enough. Here's what you suggested supporting:

```python
generator = ctranslate2.Generator("hf://ctranslate2/llama3-8b", device="cpu")
```

The equivalent currently can be as simple as this:

```python
from huggingface_hub import snapshot_download
generator = ctranslate2.Generator(snapshot_download("ctranslate2/llama3-8b", device="cpu"))
```

That's still quite simple, but it maintains a loose coupling between CTranslate2 and huggingface_hub. It has been noted that our examples need work (#1947). Our [current chat example](https://github.com/OpenNMT/CTranslate2/blob/1251f7cbbf6ee92058e8c276dc78a915e4be55dc/examples/llama3/chat.py) is 116 lines and requires learning how to convert a model before it can be used. I believe that something like the following (also [runnable via colab](https://colab.research.google.com/drive/1ZcxC6UIZyfHkG2_D6Cfwxpu8J0bo7Kmk)) would be a more helpful initial example:

```python
from huggingface_hub import snapshot_download, hf_hub_download
from tokenizers import Tokenizer
import ctranslate2

# Initialize generator and tokenizer
generator = ctranslate2.Generator(snapshot_download('jncraton/Qwen3-0.6B-ct2-int8'))
tokenizer = Tokenizer.from_file(hf_hub_download('jncraton/Qwen3-0.6B-ct2-int8', 'tokenizer.json'))

# Create Prompt and tokenize
prompt = ("<|im_start|>system\nAnswer concisely.<|im_end|>\n"
          "<|im_start|>user\nWhat is the capital of France?<|im_end|>\n"
          "<|im_start|>assistant\n"
          "<think>\n\n</think>\n\n")
prompt_tokens = tokenizer.encode(prompt).tokens

# Generate
output_tokens = generator.generate_tokens([prompt_tokens], max_length=64)
tokenizer.decode([t.token_id for t in output_tokens])
```

**Comment by BBC-Esq** [comment]
I do have a namespace on HF of "ctranslate2-4you?  ;-)

BTW I have a conversion program that I could put in the "examples" folder if that helps for the time being?

<img width="991" height="650" alt="Image" src="https://github.com/user-attachments/assets/9f2b21a0-f0ae-4503-bf93-2963487ecd89" />

**Comment by jordimas** [comment]
Hello,

> Thank you for sharing these thoughts. Something like you've described sounds like it could be very beneficial for on-boarding new users! I do have a few ideas about the path forward.
> 
> Huggingface currently lists [over 500 models that are finetunes of just Qwen 2.5 0.5b](https://huggingface.co/models?other=base_model:finetune:Qwen/Qwen2.5-0.5B). Supporting **everything compatible** would be a very large task, and someone would likely have to manage determining which models are worth fully supporting via the CTranslate2 HF organization.

Well, I was not thinking about all fine-tuned models obliviously, just the core models. For example, if you take Gemma 3, the official ones published by Google (it and non-it all sizes).

> 
> I'm not sure that we need new functionality within the CTranslate2 codebase in order to make getting started simpler. A few well-tested example models within a CTranslate2 HF organization may be enough. Here's what you suggested supporting:

Agreed, adding support for loading hf directly may be over-killing.  This provides the same benefit with a simpler approach.

**Comment by jordimas** [comment]
I created https://huggingface.co/CTranslate2HQ since https://huggingface.co/CTranslate2 was taken.

My suggestion is that interested contributors be added as admins so we can maintain a single official repository as a collective effort with the most relevant models available and properly converted. I will like add this to the documentation and the github from page.

Share your HuggingFace username and I will be adding you

**Comment by jncraton** [comment]
I'm happy to be added if that is the desired path forward. My HF username is the same as my GitHub username.

> I was not thinking about all fine-tuned models obliviously, just the core models. For example, if you take Gemma 3, the official ones published by Google (it and non-it all sizes).

That makes sense, and my fine-tune example was perhaps a little extreme, but it still leaves us to determine what counts as a "core" model. Not all new base models have new architectures. SmolLM and [SmolLM2](https://huggingface.co/HuggingFaceTB/SmolLM2-135M-Instruct/blob/main/config.json) are just Llama. [all-MiniLM-L6-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/blob/main/config.json) is just BERT. Would we quantize and upload models like those when they are released, or is the goal here just to upload a model to test and demo a new architecture? For what its worth, [onnx-community](https://huggingface.co/onnx-community) is similar to what you are proposing and currently hosts over 900 models.

**Comment by jordimas** [comment]
Hello @BBC-Esq,
Would you be interested in helping maintain a centralized HuggingFace repository for ready-to-use models? You’ve already done some great work that could be transferred over to give the project a strong starting point. Thanks!

**Comment by BBC-Esq** [comment]
@jordimas Sure, I'd be interested.  My username is "ctranslate2-4you" on HF.

**Comment by BBC-Esq** [comment]
I've seen these huggingface "spaces" as well.  I thought there was one that converted to GGUF and would tel you if a particular model isn't supported.  Perhaps we could do something similar and link to it from the HQ and GitHub repository etc?

**Comment by jncraton** [comment]
@BBC-Esq onnx-community has a space like that:

https://huggingface.co/spaces/onnx-community/convert-to-onnx

**Comment by BBC-Esq** [comment]
Does the official person want to make the page and then I'll upload the conversion script? I'll use the ONNX one as a template basically a stream lit app.

**Comment by jordimas** [comment]
Hello @BBC-Esq @jncraton. I gave you access, you should be able to create any kind of artifact. 
Proving a space to do conversions is really cool. And also have the most common models.

**Comment by BBC-Esq** [comment]
Sounds good. Did you wanna create the initial space and then one of us can work on the application that will actually do the conversion or how did you want to handle that?

**Comment by BBC-Esq** [comment]
Just uploaded the first model...I used this at the top of the "model card" in case people didn't know...causes it to show the model hierarchy, which makes it easier to find if someone is on the original Qwen3 repository looking specifically for quantized versions...

<img width="1147" height="254" alt="Image" src="https://github.com/user-attachments/assets/0a037e0d-f7b3-4851-99f4-63471dcc441e" />

**Comment by jordimas** [comment]
Hello @BBC-Esq 

> Sounds good. Did you wanna create the initial space and then one of us can work on the application that will actually do the conversion or how did you want to handle that?

I’d say that whoever has the time can go ahead. You should have permission to create the space and make any necessary changes. Since the code is open and available on HF, I think that works well too.

**Comment by jordimas** [comment]
> Just uploaded the first model...I used this at the top of the "model card" in case people didn't know...causes it to show the model hierarchy, which makes it easier to find if someone is on the original Qwen3 repository looking specifically for quantized versions...

I think that is fine to publish some initial models manually.

However, with the numbers of models that we will need to publish it will good to have an pipeline that converts the models and uploads to HF.  I think that also to have predictable name convention it is important and also some metadata in the repo with the versions of the tools, date, etc.

There are a few options but this pipeline can be stored in a space where we store the code, I will suggest to do it in Python.

**Comment by jncraton** [comment]
@BBC-Esq That looks great!

I appreciate you taking the time to properly set those metadata fields.

If we're planning to include example usage on the model cards as you've done [here](https://huggingface.co/CTranslate2HQ/Qwen3-0.6B-ct2-AWQ/blob/e2a3f771cc4049cfbab37514df59669bbdaee28e/README.md), we might consider updating the example to use `huggingface_hub` rather than a local model.

**Comment by BBC-Esq** [comment]
I've never done a "pipeline" to HF either so...this'll be new for me.  It's just a matter of time.  I could handle creating the "space" if someone else does the pipeline for example?

**Comment by BBC-Esq** [comment]
> [@BBC-Esq](https://github.com/BBC-Esq) That looks great!
> 
> I appreciate you taking the time to properly set those metadata fields.
> 
> If we're planning to include example usage on the model cards as you've done [here](https://huggingface.co/CTranslate2HQ/Qwen3-0.6B-ct2-AWQ/blob/e2a3f771cc4049cfbab37514df59669bbdaee28e/README.md), we might consider updating the example to use `huggingface_hub` rather than a local model.

True...didn't think of that...

**Comment by BBC-Esq** [comment]
I like the naming convention of

[verbatim original model name]-ct2-[AWQ/int8/float16/bfloat16 and so on...]

Anyone have any thoughts?  Thought it might be good to try and keep it uniform?

**Comment by jordimas** [comment]
> I like the naming convention of
> 
> [verbatim original model name]-ct2-[AWQ/int8/float16/bfloat16 and so on...]
> 
> Anyone have any thoughts? Thought it might be good to try and keep it uniform?

Makes sense to me. CTranslate2 performs quantization on-the-fly when loading the model. A model stored in float16 can be loaded and run as int8, int8_float16, or any other supported compute type without needing separate model files. Said this, it may be interesting to publish also int8 models because they are smaller. However, we need to explain this well then people do not get confused: you can have the model at float16 but it will be load on the most efficient type.

**Comment by BBC-Esq** [comment]
> Makes sense to me. CTranslate2 performs quantization on-the-fly when loading the model. A model stored in float16 can be loaded and run as int8, int8_float16, or any other supported compute type without needing separate model files. Said this, it may be interesting to publish also int8 models because they are smaller. However, we need to explain this well then people do not get confused: you can have the model at float16 but it will be load on the most efficient type.

Here's my position on this, but I'll defer to your because you're the primary maintainer...

First, I personally only run ctranslate2 models that fit the compute type that I want.  For example, if I'm setting it to "int8" within my script I'm not using a "float16" model because I don't like the conversion overhead.

Second, I would never run an "int8" model as "float16," for example, because there's a slight loss in quality.  When a model originally in float16 is converted to "int8" there's a loss in quality, and subsequently reconverting it to "float16" does not regain that quality, it just pads...  And correct me I'm wrong, but it's my understanding that it would use the same amount of VRAM as if I simply use a native float16 model so...there's no benefit there either.

With the whisper models on my HF account as an example... all "float32" versions were converted from the original float32 models published by OpenAI.  After they were published, OpenAI removed them and made them all float16...so I went back into the HF history and downloaded the original float32 checkpoints.  That's how much I care about even a slight loss in quality.

Thus, all the models on my HF account (e.g. float16, bfloat16, int8, etc.) are converted from the original float32 whisper models published by openai.  No up/down converting.

Moreover, when using the CPU I always use float32 since most CPUs can't run float16/bfloat16 and I want to avoid the conversion overhead.

Thus, in all my programs I have multi-factor determination of which specific whispermodel to use.

For example, if a person has a GPU without a sufficient "compute level" to run "bfloat16," it defaults to float16...unless they uncheck my "half-precision" checkbox, in which case it uses the "float32" version - i.e. it downloads a separate and discrete model.  I'm a purest in this sense I guess.

In short, I personally never liked the approach of ctranslate2 of only uploading float16 versions...that's why I converted all the permutations of whisper models for my program.  Moreover, in hindsight, I never use the "int8_float32" versions, for example, and see those permutations as confusing and not worth even benchmarking to see if there's really any benefit in VRAM or speed...

Anyways, so the issue for you to determine is if, for example, Qwen3 is one of the top notch models to include on HF...do we only upload "int8", only "bfloat16," etc.  Most models nowadays are no longer natively in float16 as when Ctranslate2 first came out; they use bfloat16.  Are we going to stick to pushing only float16 versions for the sake of compatibility with older GPU's?

I'll help upload whatever the community decides, as I have time, but food for thought!

**Comment by jordimas** [comment]
> Anyways, so the issue for you to determine is if, for example, Qwen3 is one of the top notch models to include on HF...do we only upload "int8", only "bfloat16," etc. Most models nowadays are no longer natively in float16 as when Ctranslate2 first came out; they use bfloat16. Are we going to stick to pushing only float16 versions for the sake of compatibility with older GPU's?
> 
> I'll help upload whatever the community decides, as I have time, but food for thought!

For me publishing float16 (works across all compute types with little lost) and int8 (optimized for size) it is what makes more sense but I  have no problem if we publish more models quantified in more compute type . However, consider that this is a multiplier for the number of models that we need to publish and maintain. May be we can do this for the most popular models. No big deal for me.

**Comment by jordimas** [comment]
> I've never done a "pipeline" to HF either so...this'll be new for me. It's just a matter of time. I could handle creating the "space" if someone else does the pipeline for example?

I have not tried this, but the approach is the same that I will use and looks good:

https://github.com/winstxnhdw/ct2hf

We should give it a try. If there are short comings we can always contribute to this instead of starting our effort from scratch.

**Comment by winstxnhdw** [comment]
I am happy to merge `ct2hf` to CTranslate2. But I question the value of this feature. The main value prop of CTranslate2 is the superior performance to llama.cpp during CPU inference. Having such a feature might not add much value to users. If the issue is reducing the friction of onboarding, I think writing example scripts for each model is sufficient. See https://github.com/EricLBuehler/mistral.rs/tree/master/examples

**Comment by MahmoudAshraf97** [comment]
how about we use a converter on the flight with cache? so if you load a hf model and it's not a CT2 model, it's converted first and then used, this is very similar to LLM-API in TRT-LLM where it converts the models to TRT engines under the hood and then uses them, this has larger lead time but needs no human effort to maintain a model repo, all what's needed is code maintenance which is done anyway

**Comment by a2d8a4v** [comment]
Hi, @MahmoudAshraf97 
I agree with this method. I think in recent projects such as k2 (kaldi 2.0), they've separated the [data preprocessing](https://github.com/lhotse-speech/lhotse), [training](https://github.com/k2-fsa/k2), [inferencing](https://github.com/k2-fsa/sherpa-onnx) phases into three repos, also an [example receipt](https://github.com/k2-fsa/icefall). I think it will make user easier to understand the projects, and also help projects' developers easy to maintain those repos.

**Comment by briantbutton** [comment]
For the record, I am a fan of the title goal on this issue.  Having spent a full day trying to get CTranslate2 to work on a new Ubuntu (R-Pi 5) system, I am going to take a break.  I might come back to it; CTranslate2 is supposed to be performant.  Moonshine Voice and Argos Translate were a dream to set up.

One might suspect many people getting started have a similar profile to me.  I do JavaScript, C and Bash programming, networking and audio streaming but have limited familiarity with Python and none with HuggingFace.  If somebody wants to see the pits one can fall into trying to make CTranslate2 work, I will show you.
